vyper/codegen/context.py

```diff
@@ -212,6 +215,29 @@ def deallocate_variable(self, varname, var):
 
         del self.vars[var.name]
 
+    def mark_for_deallocation(self, varname):
+        # for variables get deallocated anyway
+        if varname in self.forvars:
+            return
+        if self.vars[varname].system:
+            return
+        self._to_deallocate.add(varname)
+
+    # "mark-and-sweep", haha
+    def sweep(self):
+        tmp = set()
+        for varname in self._to_deallocate:
+            var = self.vars[varname]
+            for s in self._scopes:
+                if s not in var.blockscopes:
+                    # defer deallocation until we hit the end of its scope
```

Comment:
is this accurate? we defer the dealoc until we hit its scope again (not necessarily its end)

Reply:
or if "it" refers to the scope then still we might have to wait until multiple scopes finish

Reply:
"it" refers to the scope of the variable

Reply:
but the sweep after statements, right?

```vyper
a: uint256 = 0
for i: uint256 in range(1, 3):
  a += i # <- can't dealoc, defer
b: uint256 = 0 # a already deallocated, swept after the for
b += 1
return b  # actual scope end but a is already deallocated
```

Reply:
yea, this is the idea here